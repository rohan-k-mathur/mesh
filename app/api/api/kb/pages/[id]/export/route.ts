// app/api/kb/pages/[id]/export/route.ts
import { NextRequest, NextResponse } from 'next/server';
import { prisma } from '@/lib/prismaclient';
import { requireKbRole, fail } from '@/lib/kb/withSpaceAuth';

export const dynamic = 'force-dynamic'; export const revalidate = 0;

function mdEscape(s: string) { return s.replace(/[*_`]/g, m => '\\'+m); }

export async function GET(req: NextRequest, { params }: { params: { id: string } }) {
  try {
    await requireKbRole(req, { pageId: params.id, need: 'reader' });

    const url = new URL(req.url);
    const as = (url.searchParams.get('as') ?? 'md').toLowerCase();
    const at = url.searchParams.get('at'); // 'latest' | ISO (Phase‑B can support snapshot ids)

    const page = await prisma.kbPage.findUnique({
      where: { id: params.id },
      select: { id:true, title:true, summary:true, tags:true, spaceId:true, frontmatter:true, blocks: {
        orderBy: [{ ord: 'asc' }, { createdAt: 'asc' }],
        select: { id:true, type:true, ord:true, live:true, dataJson:true, pinnedJson:true }
      }}
    });
    if (!page) return NextResponse.json({ error: 'not_found' }, { status: 404 });

    // Manifest (pinned later; for now we only record what's live vs pinned)
    const manifest = {
      page: { id: page.id, title: page.title, at: at ?? 'latest' },
      blocks: page.blocks.map(b => ({
        id: b.id, type: b.type, live: b.live, pinned: !b.live && !!b.pinnedJson
      }))
    };

    // Very light serializer (Phase‑A): Structured blocks → short sections with deep links.
    const lines: string[] = [];
    lines.push('---');
    lines.push(`title: ${mdEscape(page.title)}`);
    if (page.summary) lines.push(`summary: ${mdEscape(page.summary)}`);
    if (page.tags?.length) lines.push(`tags: [${page.tags.map(mdEscape).join(', ')}]`);
    if (page.frontmatter) lines.push(`frontmatter: ${JSON.stringify(page.frontmatter)}`);
    lines.push('---\n');

    lines.push(`# ${page.title}\n`);
    if (page.summary) lines.push(`${page.summary}\n`);

    for (const b of page.blocks) {
      const data: any = b.live ? b.dataJson : (b.pinnedJson ?? b.dataJson);

      if (b.type === 'text') {
        const md = (data?.md as string) ?? '';
        lines.push(md.trim()); lines.push('');
        continue;
      }
      if (b.type === 'claim') {
        lines.push(`> **Claim** \`${data?.id ?? '—'}\``);
        lines.push(`> lens: ${data?.lens ?? 'belpl'} · room: ${data?.roomId ?? 'auto'}`);
        lines.push('');
        continue;
      }
      if (b.type === 'argument') {
        lines.push(`> **Argument** \`${data?.id ?? '—'}\` (diagram)`);
        lines.push('');
        continue;
      }
      if (b.type === 'room_summary') {
        lines.push(`> **Room summary** \`${data?.id ?? '—'}\` limit=${data?.limit ?? 5}`);
        lines.push('');
        continue;
      }
      if (b.type === 'sheet') {
        lines.push(`> **Sheet** \`${data?.id ?? '—'}\``);
        lines.push('');
        continue;
      }
      if (b.type === 'transport') {
        lines.push(`> **Transport** ${data?.fromId ?? '—'} → ${data?.toId ?? '—'}`);
        lines.push('');
        continue;
      }
      // Fallback
      lines.push(`> **${b.type}** (unsupported in md export)`); lines.push('');
    }

    // Embed manifest for reproducibility
    lines.push('```json x-kb-manifest');
    lines.push(JSON.stringify(manifest, null, 2));
    lines.push('```');

    if (as === 'md') {
      const md = lines.join('\n');
      return new NextResponse(md, {
        status: 200,
        headers: {
          'Content-Type': 'text/markdown; charset=utf-8',
          'Cache-Control': 'no-store',
          'Content-Disposition': `attachment; filename="${page.title.replace(/\s+/g,'_')}.md"`
        }
      });
    }

    if (as === 'pdf') {
      // Stub: return print‑ready HTML (Phase‑B: render via headless/puppeteer or a PDF service)
      const html = `<!doctype html><meta charset="utf-8">
        <title>${page.title}</title>
        <style>body{font:14px/1.5 system-ui, -apple-system, Segoe UI, Roboto, sans-serif; max-width:720px; margin:40px auto; color:#0f172a}
        h1{font-size:28px;margin:0 0 12px} blockquote{border-left:3px solid #cbd5e1; margin:12px 0; padding:6px 10px; color:#334155}
        pre{background:#f8fafc;border:1px solid #e2e8f0;padding:12px;border-radius:6px;overflow:auto}</style>
        <h1>${page.title}</h1>
        ${lines.filter(l=>!l.startsWith('```')).map(l => l.startsWith('> ') ? `<blockquote>${l.slice(2)}</blockquote>` : `<p>${l.replace(/^#\s+.*$/,'')}</p>`).join('\n')}
      `;
      return new NextResponse(html, { status: 200, headers: { 'Content-Type': 'text/html; charset=utf-8' } });
    }

    return NextResponse.json({ error: 'unsupported_format' }, { status: 400 });
  } catch (e) { return fail(e); }
}
