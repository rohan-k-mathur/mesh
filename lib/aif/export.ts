// //lib/aif/export.ts
// import ctx from './context.json';
// import { prisma } from '@/lib/prismaclient';

// type Json = Record<string, any>;

// function N(type: string, id: string, extra: Json = {}) {
//   return { '@id': id, '@type': type, ...extra };
// }
// function E(from: string, to: string, role: string) {
//   return { '@id': `:e_${from.replace(':','|')}_${role}_${to.replace(':','|')}`, '@type': 'aif:Edge', role, from, to };
// }

// const idI  = (cid: string) => `:I|${cid}`;
// const idRA = (aid: string) => `:RA|${aid}`;
// const idCA = (cid: string) => `:CA|${cid}`;
// const idPA = (pid: string) => `:PA|${pid}`;
// const idPM = (wid: string) => `:PM|${wid}`;      // Pascal meta bundle per Work(=theoryWork)


// export async function exportDeliberationAsAifJSONLD(deliberationId: string) {
  
//   const [ moves] = await Promise.all([
//     prisma.claim.findMany({ where: { deliberationId } }),
//     prisma.argument.findMany({
//       where: { deliberationId },
//       include: { scheme: true, premises: true, conclusion: true }
//     }),
//     prisma.argumentEdge.findMany({ where: { deliberationId } }),
//     prisma.dialogueMove.findMany({ where: { deliberationId } })
//   ]);

// function n(type: string, id: string, extra: any = {}) {
//   return { '@id': `:${id}`, '@type': type, ...extra };
// }
// function edge(from: string, to: string, role: 'premise'|'conclusion'|'conflictingElement'|'conflictedElement'|'preferredElement'|'dispreferredElement') {
//   return { '@id': `:e_${from}_${role}_${to}`, '@type': 'aif:Edge', role, from: `:${from}`, to: `:${to}` };
// }
//   const N: any[] = [];
//   const E: any[] = [];


// const claims = await prisma.claim.findMany({ where:{ deliberationId }, select:{ id:true, text:true } });
// const args   = await prisma.argument.findMany({ where:{ deliberationId }, include:{ premises:{ select:{ claimId:true } }, conclusion:true, scheme:true } });
// const cas    = await prisma.conflictApplication.findMany({ where:{ deliberationId }, include:{ scheme:true } });
// const pas    = await prisma.preferenceApplication.findMany({ where:{ deliberationId }, include:{ scheme:true } });

// const idI  = (cid:string)=>`I|${cid}`;
// const idRA = (aid:string)=>`RA|${aid}`;
// const idCA = (cid:string)=>`CA|${cid}`;
// const idPA = (pid:string)=>`PA|${pid}`;

// const nodes:any[] = [];
// const edges:any[] = [];


// // const nodes:any[] = [];
// // const edges:any[] = [];

// // // I-nodes (Claims)
// // for (const c of claims) nodes.push(n('aif:InformationNode', `I_${c.id}`, { text: c.text ?? '' }));

// // // RA-nodes (Arguments)
// // for (const a of args) nodes.push(n('aif:RA', `RA_${a.id}`, { schemeKey: a.schemeId ?? null }));

// // // Premises & conclusion edges
// // for (const a of args) {
// //   for (const p of a.premises) edges.push(edge(`I_${p.claimId}`, `RA_${a.id}`, 'premise'));
// //   edges.push(edge(`RA_${a.id}`, `I_${a.conclusionClaimId}`, 'conclusion'));
// // }

// // CA-nodes
// for (const ca of conflictApps) {
//   const caId = `CA_${ca.id}`; nodes.push(n('aif:CA', caId, { schemeKey: ca.schemeKey ?? null }));
//   const confFrom = ca.conflictingArgumentId ? `RA_${ca.conflictingArgumentId}` : `I_${ca.conflictingClaimId}`;
//   const confTo   = ca.conflictedArgumentId  ? `RA_${ca.conflictedArgumentId}`  : `I_${ca.conflictedClaimId}`;
//   edges.push(edge(confFrom, caId, 'conflictingElement'));
//   edges.push(edge(caId, confTo, 'conflictedElement'));
// }

// // PA-nodes
// for (const pa of preferenceApps) {
//   const paId = `PA_${pa.id}`; nodes.push(n('aif:PA', paId, { schemeKey: pa.schemeKey ?? null }));
//   const prefFrom = pa.preferredArgumentId ? `RA_${pa.preferredArgumentId}` : `I_${pa.preferredClaimId}`;
//   const prefTo   = pa.dispreferredArgumentId ? `RA_${pa.dispreferredArgumentId}` : `I_${pa.dispreferredClaimId}`;
//   edges.push(edge(prefFrom, paId, 'preferredElement'));
//   edges.push(edge(paId, prefTo, 'dispreferredElement'));
// }
  
//   // I-nodes
//   for (const c of claims) {
//     N.push({ "@id": `I:${c.id}`, "@type": "aif:InformationNode", text: c.text });
//   }

//   // RA-nodes
//   for (const a of args) {
//     N.push({
//       "@id": `S:${a.id}`,
//       "@type": "aif:RA",
//       "usesScheme": a.scheme?.key ?? null
//     });
//     for (const p of a.premises)
//       E.push({ from: `I:${p.claimId}`, to: `S:${a.id}`, role: "aif:Premise" });

//     E.push({ from: `S:${a.id}`, to: `I:${a.conclusionClaimId}`, role: "aif:Conclusion" });
//   }

//   // CA-nodes (one CA node per edge)
//   for (const e of edges) {
//     const caId = `CA:${e.id}`;
//     N.push({
//       "@id": caId, "@type": "aif:CA",
//       "attackType": e.attackType, "targetScope": e.targetScope, "cqKey": e.cqKey ?? null
//     });
//     // premise into CA is the conclusion of the attacking RA
//     const attacker = args.find(a => a.id === e.fromArgumentId)!;
//     E.push({ from: `I:${attacker.conclusionClaimId}`, to: caId, role: "aif:Premise" });

//     if (e.targetScope === 'conclusion' && e.targetClaimId) {
//       E.push({ from: caId, to: `I:${e.targetClaimId}`, role: "aif:Attacks" });
//     } else if (e.targetScope === 'inference' && e.toArgumentId) {
//       E.push({ from: caId, to: `S:${e.toArgumentId}`, role: "aif:Attacks" });
//     } else if (e.targetScope === 'premise' && e.targetPremiseId) {
//       E.push({ from: caId, to: `I:${e.targetPremiseId}`, role: "aif:Attacks" });
//     }
//   }

//   // L-nodes (locutions)
//   for (const m of moves) {
//     N.push({
//       "@id": `L:${m.id}`,
//       "@type": "aif:L",
//       "illocution": m.illocution,
//       "text": m.payload?.expression ?? null
//     });
//     if (m.argumentId) {
//       E.push({ from: `L:${m.id}`, to: `S:${m.argumentId}`, role: "aif:Illocutes" });
//     } else if (m.contentClaimId) {
//       E.push({ from: `L:${m.id}`, to: `I:${m.contentClaimId}`, role: "aif:Illocutes" });
//     }
//     if (m.replyToMoveId) {
//       E.push({ from: `L:${m.replyToMoveId}`, to: `L:${m.id}`, role: "aif:Replies" });
//     }
//   }

// //   return { "@context": ctx["@context"], nodes: N, edges: E };

// // // I-nodes
// // for (const c of claims) nodes.push({ '@id': idI(c.id), '@type':'aif:InformationNode', text: c.text ?? '' });

// // // RA-nodes + role edges
// // for (const a of args) {
// //   nodes.push({ '@id': idRA(a.id), '@type':'aif:RA', scheme: a.scheme?.key ?? null });
// //   for (const p of a.premises) edges.push({ '@type':'aif:Premise',    from: idI(p.claimId), to: idRA(a.id) });
// //   if (a.conclusionClaimId)   edges.push({ '@type':'aif:Conclusion', from: idRA(a.id),     to: idI(a.conclusionClaimId) });
// // }

// // // CA-nodes + role edges
// // for (const ca of cas) {
// //   nodes.push({ '@id': idCA(ca.id), '@type':'aif:CA', scheme: ca.scheme?.key ?? null });
// //   const conflicting = ca.conflictingArgumentId ? idRA(ca.conflictingArgumentId) : idI(ca.conflictingClaimId!);
// //   const conflicted  = ca.conflictedArgumentId  ? idRA(ca.conflictedArgumentId)  : idI(ca.conflictedClaimId!);
// //   // NB: S→S allowed; I→I disallowed except via this CA node (we comply)
// //   edges.push({ '@type':'aif:ConflictingElement', from: conflicting, to: idCA(ca.id) });
// //   edges.push({ '@type':'aif:ConflictedElement',  from: idCA(ca.id), to:  conflicted });
// // }

// // // PA-nodes + role edges
// // const ref = (cid?:string, aid?:string, sid?:string) => aid ? idRA(aid) : (cid ? idI(cid) : (sid ? `SCH|${sid}` : ''));
// // for (const pa of pas) {
// //   nodes.push({ '@id': idPA(pa.id), '@type':'aif:PA', scheme: pa.scheme?.key ?? null });
// //   const pref = ref(pa.preferredClaimId, pa.preferredArgumentId, pa.preferredSchemeId);
// //   const disp = ref(pa.dispreferredClaimId, pa.dispreferredArgumentId, pa.dispreferredSchemeId);
// //   edges.push({ '@type':'aif:PreferredElement',    from: pref,   to: idPA(pa.id) });
// //   edges.push({ '@type':'aif:DispreferredElement', from: idPA(pa.id), to: disp   });
// // }


//   // Optional: AIF+ locutions (minimal)
//   for (const m of moves) {
//     const L = `:L|${m.id}`;
//     nodes.push(N('aif:L', L, { illocution: m.illocution ?? null, text: m.payload?.expression ?? null }));
//     if (m.argumentId)      edges.push(E(L, idRA(m.argumentId), 'aif:Illocutes'));
//     else if (m.contentClaimId) edges.push(E(L, idI(m.contentClaimId), 'aif:Illocutes'));
//     if (m.replyToMoveId)   edges.push(E(`:L|${m.replyToMoveId}`, L, 'aif:Replies'));
//   }

//   // OP/Pascal preconditions bundle (if any OP work w/ pascal model exists)
//   const opWithPascal = works.find(w => !!w.pascal);
//   if (opWithPascal?.pascal) {
//     const { propositions, actions, utilities, method, assumption } = opWithPascal.pascal as any;
//     nodes.push(N('aif:PascalMeta', idPM(opWithPascal.id), {
//       workId: opWithPascal.id,
//       method,                                       // 'laplace' | 'minimax' | 'regret'
//       assumption: assumption ?? null,               // TOP2/TOP3 support text
//       propositions, actions, utilities              // matrix to reconstruct expected desirability
//     }));
//     // Link PascalMeta to any OP claims (optional; safe no-op if none)
//     for (const rc of await prisma.theoryWorkClaim.findMany({ where: { workId: opWithPascal.id } })) {
//       edges.push(E(idPM(opWithPascal.id), idI(rc.claimId), 'aif:Annotates'));
//     }
//   }

//   return {
//     "@context": ctx["@context"],
//     "nodes": nodes,
//     "edges": edges
//   };
// }
// lib/aif/export.ts
import ctx from './context.json';
import { prisma } from '@/lib/prismaclient';

type Json = Record<string, any>;

function N(type: string, id: string, extra: Json = {}) {
  return { '@id': id, '@type': type, ...extra };
}
function E(from: string, to: string, role: string) {
  return { '@id': `:e_${from.replace(':','|')}_${role}_${to.replace(':','|')}`, '@type': 'aif:Edge', role, from, to };
}

const idI  = (cid: string) => `:I|${cid}`;
const idRA = (aid: string) => `:RA|${aid}`;
const idCA = (cid: string) => `:CA|${cid}`;
const idPA = (pid: string) => `:PA|${pid}`;
const idPM = (wid: string) => `:PM|${wid}`;      // Pascal meta bundle per Work(=theoryWork)

export async function exportDeliberationAsAifJSONLD(deliberationId: string) {
  // Pull graph primitives (+ optional preferences if you have them)
  const [claims, args, cas, pas, moves, works] = await Promise.all([
    prisma.claim.findMany({ where:{ deliberationId }, select:{ id:true, text:true } }),
    prisma.argument.findMany({
      where:{ deliberationId },
      include:{ premises:{ select:{ claimId:true } }, conclusion:true, scheme:true }
    }),
    prisma.conflictApplication.findMany({ where:{ deliberationId }, include:{ scheme:true } }),
    prisma.preferenceApplication.findMany({ where:{ deliberationId }, include:{ scheme:true } }),
    prisma.dialogueMove.findMany({ where:{ deliberationId } }),
    prisma.theoryWork.findMany({
      where: { deliberationId, theoryType: 'OP' },
      include: { pascal: true }
    })
  ]);

  const nodes: any[] = [];
  const edges: any[] = [];

  // I-nodes
  for (const c of claims) nodes.push(N('aif:InformationNode', idI(c.id), { text: c.text ?? '' }));

  // RA-nodes + premise/conclusion edges
  for (const a of args) {
    nodes.push(N('aif:RA', idRA(a.id), { schemeKey: a.scheme?.key ?? null }));
    for (const p of a.premises) edges.push(E(idI(p.claimId), idRA(a.id), 'aif:Premise'));
    if (a.conclusionClaimId) edges.push(E(idRA(a.id), idI(a.conclusionClaimId), 'aif:Conclusion'));
  }

  // CA-nodes (conflicts)
  for (const ca of cas) {
    const thisCA = idCA(ca.id);
    nodes.push(N('aif:CA', thisCA, { schemeKey: ca.scheme?.key ?? null, attackType: ca.attackType ?? null, targetScope: ca.targetScope ?? null }));
    const conflicting = ca.conflictingArgumentId ? idRA(ca.conflictingArgumentId) : idI(ca.conflictingClaimId!);
    const conflicted  = ca.conflictedArgumentId  ? idRA(ca.conflictedArgumentId)  : idI(ca.conflictedClaimId!);
    edges.push(E(conflicting, thisCA, 'aif:ConflictingElement'));
    edges.push(E(thisCA,      conflicted, 'aif:ConflictedElement'));
  }

  // PA-nodes (preferences)
  for (const pa of pas) {
    const thisPA = idPA(pa.id);
    nodes.push(N('aif:PA', thisPA, { schemeKey: pa.scheme?.key ?? null }));
    const preferred     = pa.preferredArgumentId     ? idRA(pa.preferredArgumentId)     : idI(pa.preferredClaimId!);
    const dispreferred  = pa.dispreferredArgumentId  ? idRA(pa.dispreferredArgumentId)  : idI(pa.dispreferredClaimId!);
    edges.push(E(preferred,    thisPA, 'aif:PreferredElement'));
    edges.push(E(thisPA, dispreferred, 'aif:DispreferredElement'));
  }

  // Optional: AIF+ locutions (minimal)
  for (const m of moves) {
    const L = `:L|${m.id}`;
    nodes.push(N('aif:L', L, { illocution: m.illocution ?? null, text: m.payload?.expression ?? null }));
    if (m.argumentId)      edges.push(E(L, idRA(m.argumentId), 'aif:Illocutes'));
    else if (m.contentClaimId) edges.push(E(L, idI(m.contentClaimId), 'aif:Illocutes'));
    if (m.replyToMoveId)   edges.push(E(`:L|${m.replyToMoveId}`, L, 'aif:Replies'));
  }

  // OP/Pascal preconditions bundle (if any OP work w/ pascal model exists)
  const opWithPascal = works.find(w => !!w.pascal);
  if (opWithPascal?.pascal) {
    const { propositions, actions, utilities, method, assumption } = opWithPascal.pascal as any;
    nodes.push(N('aif:PascalMeta', idPM(opWithPascal.id), {
      workId: opWithPascal.id,
      method,                                       // 'laplace' | 'minimax' | 'regret'
      assumption: assumption ?? null,               // TOP2/TOP3 support text
      propositions, actions, utilities              // matrix to reconstruct expected desirability
    }));
    // Link PascalMeta to any OP claims (optional; safe no-op if none)
    for (const rc of await prisma.theoryWorkClaim.findMany({ where: { workId: opWithPascal.id } })) {
      edges.push(E(idPM(opWithPascal.id), idI(rc.claimId), 'aif:Annotates'));
    }
  }

  return {
    "@context": ctx["@context"],
    "nodes": nodes,
    "edges": edges
  };
}
