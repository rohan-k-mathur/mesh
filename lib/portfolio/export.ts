import { escapeHtml, escapeJSX, camelToKebab } from "@/lib/utils/escape";


export interface PortfolioExportData {
  text: string;
  images: string[];
  links: string[];
  layout: "grid" | "column" | "free";
  color: string;
  absolutes?: AbsoluteElement[];


  snapshot?: string;          // public URL of PNG
  snapshotWidth?: number;     // optional natural size
  snapshotHeight?: number;
}
export interface AbsoluteElement {
  id: string;
  type: "text" | "text-box" | "image" | "link" | "box";
  x: number;
  y: number;
  width: number;
  height: number;
  natW?: number;
  natH?: number;
  content?: string;
  src?: string;
  href?: string;
}
export function generatePortfolioTemplates(
  data: PortfolioExportData,
): { html: string; css: string; tsx?: string } {
  /* ❷ Free‑layout → React branch */
  if (data.layout === "free" && data.absolutes?.length) {
    const bg = data.color || "bg-white";

    /* 📏 overall canvas bounds */
    const canvasWidth  = Math.max(...data.absolutes.map(a => a.x + a.width))  + 20;
    const canvasHeight = Math.max(...data.absolutes.map(a => a.y + a.height)) + 20;

    /* ---------- 1‑A  plain HTML fallback ---------- */
    const itemsHtml = data.absolutes
      .map(a => {
        const style = `style="position:absolute;left:${a.x}px;top:${a.y}px;width:${a.width}px;height:${a.height}px;object-fit:cover"`;
        switch (a.type) {
          case "image":
            return `<img ${style} src="${escapeHtml(a.src)}" />`;
          case "link":
            return `<a ${style} href="${escapeHtml(a.href)}" class="break-all underline text-blue-500" target="_blank" rel="noreferrer">${escapeHtml(a.href)}</a>`;
          default: // "text" | "text-box" | "box"
            return `<div ${style} class="text-xs border p-1 overflow-hidden">${escapeHtml(
              a.content,
            )}</div>`;
        }
      })
      .join("\n");

    const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <title>Portfolio</title>
  <link href="./tailwind.css" rel="stylesheet" />
</head>
<body class="${bg} flex items-center justify-center min-h-screen">
  <div style="position:relative;width:${canvasWidth}px;height:${canvasHeight}px">
    ${itemsHtml}
  </div>
</body>
</html>`;

    /* ---------- 1‑B  React component string (.tsx) ---------- */
    const reactBody = data.absolutes
      .map(a => {
        const styleObj = {
          position: "absolute",
          left: a.x,
          top: a.y,
          width: a.width,
          height: a.height,
        };
        const styleJsx = `{ ${Object.entries(styleObj)
          .map(([k, v]) => `${k}: ${v}`)
          .join(", ")} }`;

        switch (a.type) {
          case "image":
            return `<Image src="${escapeJSX(a.src)}" alt="" width={${a.width}} height={${a.height}} style={{ position: 'absolute', left: ${a.x}, top: ${a.y}, width: ${a.width}, height: ${a.height}, objectFit: 'cover' }} />`;
          case "link":
            return `<a href="${escapeJSX(a.href)}" style={${styleJsx}} className="break-all underline text-blue-500" target="_blank" rel="noreferrer">${escapeJSX(
              a.href,
            )}</a>`;
          case "box":
            return `<div style={${styleJsx}} className="bg-gray-200 border" />`;
          default: // "text" | "text-box"
            return `<div style={${styleJsx}} className="text-xs border p-1 overflow-hidden whitespace-pre-wrap">${escapeJSX(
              a.content,
            )}</div>`;
        }
      })
      .join("\n");

    const tsx = `/* auto‑generated */
import Image from "next/image";

export default function PortfolioCanvas() {
  return (
    <div
      className="${bg} flex items-center justify-center min-h-screen"
      style={{ overflow: "auto" }}
    >
      <div style={{
        position: "relative",
        width: ${canvasWidth},
        height: ${canvasHeight},
        }}>
        ${reactBody}
      </div>
    </div>
  );
}
`;

    const css = `@tailwind base;\n@tailwind components;\n@tailwind utilities;`;
    return { html, css, tsx };
  }
  
  if (data.layout!="free" && data.snapshot) {
    const w = data.snapshotWidth ?? 1200;
    const h = data.snapshotHeight ?? 800;
    const bgClass = data.color || "bg-white";

    const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <title>Portfolio</title>
  <link href="./tailwind.css" rel="stylesheet" />
  <meta property="og:image" content="${data.snapshot}" />
</head>
<body class="${bgClass} flex items-center justify-center min-h-screen">
  <img src="${data.snapshot}" alt="Portfolio snapshot" width="${w}" height="${h}"
       class="max-w-full h-auto shadow-lg rounded" />
</body>
</html>`;

    /* no custom CSS needed, but keep empty shell so downstream code stays same */
    const css = `@tailwind base;\n@tailwind components;\n@tailwind utilities;`;
    return { html, css };
  }


  const bgClass = data.color || "bg-white";

  const textBlock = data.text
    ? `<div class="text-block flex flex-col max-h-[1000px] mb-1 mt-2 break-words">${data.text}</div>`
    : "";

  const absImgs = data.absolutes?.filter(a => a.type === "image") ?? [];
  const images = (absImgs.length ? absImgs : data.images.map(src => ({ src })))
    .map((img, idx) => {
      const src = (img as any).src ?? img;
      const w = (img as any).width;
      const h = (img as any).height;
      const style = w && h
        ? `style="width:${w}px;height:${h}px;object-fit:cover"`
        : "";
      return `<img src="${src}" ${style} class="portfolio-img-frame ${idx === 0 ? "flex flex-col max-h-[3000px] mb-1 mt-2 break-words" : ""}" />`;
    })
    .join("\n");

  const links = data.links
    .map(
      (href) =>
        `<a href="${href}" class="text-blue-500 underline break-all" target="_blank" rel="noreferrer">${href}</a>`
    )
    .join("\n");

  const layoutClass = data.layout === "grid" ? "grid grid-cols-2 gap-2" : "flex flex-col gap-2";

  const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Portfolio</title>
  <link href="./tailwind.css" rel="stylesheet">
</head>
<body>
  <div class="portfolio-container flex flex-col w-[34rem] max-h-[3000px]">
    <div class="${layoutClass} rounded-lg p-4 ${bgClass} mt-4 max-h-[3000px]">
      ${textBlock}
      ${images}
      ${links}
    </div>
  </div>
</body>
</html>`;

  const css = `@tailwind base;\n@tailwind components;\n@tailwind utilities;\n\n.portfolio-container {\n  display: flex;\n  padding: 50px;\n  width: 34rem;\n}\n\n.portfolio-img-frame {\n  background: transparent;\n  border: 1px solid #1a192b;\n  padding: 0px;\n}\n\n.text-block {\n  width: fit-content;\n  border: .5px solid #000000;\n  border-radius: 2px;\n  font-size: .7rem;\n  line-height: 1;\n  letter-spacing: normal;\n  margin: 0.3rem 0;\n  padding: 0.3rem;\n  max-height: 9.25rem;\n  overflow-y: auto;\n  scroll-behavior: smooth;\n}\n\n.grid {\n  display: grid;\n  grid-template-columns: repeat(2, minmax(0, 1fr));\n  gap: 0.5rem;\n}\n.flex {\n  display: flex;\n}\n`;

   return { html, css };
}


// const reactCode = `import Image from "next/image";
// export default function PortfolioCanvas() {
//   return (
//     <div style={{ position: "relative", width: ${canvasWidth}, height: ${canvasHeight} }}>
//       ${data.absolutes
//         .map(a => {
//           const style = `{ position:"absolute", left:${a.x}, top:${a.y},
//                            width:${a.width}, height:${a.height} }`;
//           switch (a.type) {
//             case "image":
//               return `<Image src="${a.src}" alt="" fill style=${style} />`;
//             case "text":
//             case "text-box":
//               return `<div style=${style} className="text-xs border p-1">${escapeJSX(
//                 a.content || "",
//               )}</div>`;
//             // etc.
//           }
//         })
//         .join("\n")}
//     </div>
//   );
// }`;