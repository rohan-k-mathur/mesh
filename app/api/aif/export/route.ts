import { NextRequest, NextResponse } from 'next/server';
import { buildAifGraphJSONLD } from '@/lib/aif/jsonld';

export const dynamic = 'force-dynamic';
export const revalidate = 0;

export async function GET(req: NextRequest) {
  const u = new URL(req.url);
  const deliberationId = u.searchParams.get('deliberationId') ?? undefined;
  const includeLocutions = u.searchParams.get('loc') === '1';
  const includeCQs = u.searchParams.get('cqs') === '1';

  if (!deliberationId) {
    return NextResponse.json({ ok:false, error: 'missing deliberationId' }, { status: 400 });
  }

  const doc = await buildAifGraphJSONLD({ deliberationId, includeLocutions, includeCQs });
  return NextResponse.json({ ok:true, ...doc }, { headers: { 'Cache-Control': 'no-store' } });
}
