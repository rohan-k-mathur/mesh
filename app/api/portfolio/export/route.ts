import { NextRequest, NextResponse } from "next/server";
import { generatePortfolioTemplates, PortfolioExportData } from "@/lib/portfolio/export";
import { createPortfolioPage } from "@/lib/actions/portfolio.actions";
import { screenshotPage } from "@/lib/screenshot";

import { getUserFromCookies } from "@/lib/serverutils";
import { createRealtimePost } from "@/lib/actions/realtimepost.actions";
import { uploadFileToSupabase } from "@/lib/utils";
export const runtime = "nodejs";       // make sure Playwright can run

// const { html, css, tsx } = generatePortfolioTemplates(data);

// await createPortfolioPage({
//   html,
//   css,
//   tsx,          // NEW (can be undefined)
// });


export async function POST(req: NextRequest) {
  const data = (await req.json()) as PortfolioExportData;
  // const user  = await getUserFromCookies();   // safe here



  const { html, css, tsx } = generatePortfolioTemplates(data);

  const slug = await createPortfolioPage({ html, css, tsx });

    /* 2) render â†’ screenshot */
    const base = process.env.NEXT_PUBLIC_BASE_URL || "http://localhost:3000";
    const pageUrl = `${base}/portfolio/${slug}`;

    const pngBuffer = await screenshotPage(pageUrl);
  const fileName  = `portfolio/snap-${slug}.png`;
  const { fileURL: snapshot, error } = await uploadFileToSupabase(
    new File([pngBuffer], fileName, { type: "image/png" })
  );
  if (error) throw error;

  /* 3) respond with everything the client needs */
  return NextResponse.json({
    url: `/portfolio/${slug}`,
    snapshot,
    snapshotWidth: 800,   // match viewport you passed to screenshotPage
    snapshotHeight: 1200,
  });
}
