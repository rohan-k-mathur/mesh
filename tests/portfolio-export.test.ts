import { generatePortfolioTemplates } from "@/lib/portfolio/export";

test("portfolio export matches snapshot", () => {
  const data = {
    text: "Hello",
    images: ["img.png"],
    links: ["https://example.com"],
    layout: "column" as const,
    color: "bg-white",
  };

  const { html, css } = generatePortfolioTemplates(data);
  expect(html).toMatchSnapshot();
  expect(css).toMatchSnapshot();
});
