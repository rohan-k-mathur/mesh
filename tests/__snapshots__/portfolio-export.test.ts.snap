// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`portfolio export matches snapshot 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Portfolio</title>
  <link href="./tailwind.css" rel="stylesheet">
</head>
<body>
  <div class="portfolio-container flex flex-col w-[34rem] max-h-[3000px]">
    <div class="flex flex-col gap-2 rounded-lg p-4 bg-white mt-4 max-h-[3000px]">
      <div class="text-block flex flex-col max-h-[1000px] mb-1 mt-2 break-words">Hello</div>
      <img src="img.png"  class="portfolio-img-frame flex flex-col max-h-[3000px] mb-1 mt-2 break-words" />
      <a href="https://example.com" class="text-blue-500 underline break-all" target="_blank" rel="noreferrer">https://example.com</a>
    </div>
  </div>
</body>
</html>"
`;

exports[`portfolio export matches snapshot 2`] = `
"@tailwind base;
@tailwind components;
@tailwind utilities;

.portfolio-container {
  display: flex;
  padding: 50px;
  width: 34rem;
}

.portfolio-img-frame {
  background: transparent;
  border: 1px solid #1a192b;
  padding: 0px;
}

.text-block {
  width: fit-content;
  border: .5px solid #000000;
  border-radius: 2px;
  font-size: .7rem;
  line-height: 1;
  letter-spacing: normal;
  margin: 0.3rem 0;
  padding: 0.3rem;
  max-height: 9.25rem;
  overflow-y: auto;
  scroll-behavior: smooth;
}

.grid {
  display: grid;
  grid-template-columns: repeat(2, minmax(0, 1fr));
  gap: 0.5rem;
}
.flex {
  display: flex;
}
"
`;
